python/rmm/_lib/memory_resource.pyx

```diff
@@ -444,6 +464,62 @@ cdef class BinningMemoryResource(UpstreamResourceAdaptor):
                     bin_resource.get_mr())
 
 
+cdef void* _allocate_callback_wrapper(
+    size_t nbytes,
+    cuda_stream_view stream,
+    void *ctx
+) with gil:
+    return <void*><uintptr_t>((<object>ctx)(nbytes))
```

Comment:
Is it meaningful to cast this twice?

Reply:
Yes, because the returned value of `(<object>ctx)(nbytes)` is a Python int, so it needs to be cast to a C type first.

Reply:
I agree that it's meaningful, but is it necessary? This is a `cdef` function with a `void *` return type. Won't the generated code automatically perform the cast? We rely on that in the corresponding `deallocate` wrapper below, the first parameter is passed as a `uintptr_t` even though it should be a `void *`.

Reply:
It is necessary, otherwise we end up casting the `PyObject*` to a `void*`, rather than casting to an `intptr_t` first.

> We rely on that in the corresponding deallocate wrapper below, the first parameter is passed as a uintptr_t even though it should be a void *.

Sorry, I'm not sure I follow -- the `deallocate_wrapper` is invoked in C++ land. The `void*` that it is passed is cast into a `uintptr_t` and that `uintptr_t` is passed to the user-provided Python callback for deallocation.

Reply:
Oh apparently @shwina and I were running in parallel without a lock...

Reply:
Sorry, my statement was obviously pretty ambiguous here. I wasn't not suggesting
```
<void*>((<object>ctx)(nbytes))
```
which is obviously a recipe for disaster. I was asking whether we could instead do this
```
<uintptr_t>((<object>ctx)(nbytes))
```
and rely on Cython to cast the `uintptr_t` to a `void*` because it looked like that is what was happening inside `_deallocate_callback_wrapper` where it's doing
```
(<object>ctx)(<uintptr_t>(ptr), nbytes)
```
rather than
```
(<object>ctx)(<void*><uintptr_t>(ptr), nbytes)
```
In C these types won't automatically decay to one another, but I thought Cython might have added some casting rules. 

I wasn't suggesting that the auto cast was _better_, just that we should be consistent between these two usages. However, on a closer read I'm no longer sure that this is an apples to apples comparison because `ctx` here will be a Python function, so maybe it is necessary for the parameter in `_deallocate_calback_wrapper` to be a `uintptr_t` that is only cast to `void*` inside the Cython `deallocate` call [on line 189 above](https://github.com/rapidsai/rmm/pull/980/files#diff-13a7108445229590ccb91d6b9eda5334d03cd22c3d2379cd09e7c9bed58deafdR189).

Reply:
Auto-casting does not work here:
```
Error compiling Cython file:
------------------------------------------------------------
...
from libc.stdint cimport intptr_t


cdef void* my_cast(x):
    return <intptr_t>(x)
          ^
------------------------------------------------------------

test_cython_cast.pyx:5:11: Cannot assign type 'intptr_t' to 'void *'
```

Reply:
Makes sense. When I first commented I didn't realize that the parameter in the `_deallocate_callback_wrapper` is in fact eventually manually cast to `void*` inside `DeviceMemoryResource.deallocate`, that's what I was missing here. Thanks for the fact-checking.